automata/base/automaton.py

```diff
@@ -78,15 +78,16 @@ def _validate_final_states(self):
                 'final states are not valid ({})'.format(
                     ', '.join(str(state) for state in invalid_states)))
 
-    def attributes(self):
+    @property
```

Comment:
Isn't there a way to disable the setter if you're declaring things with the property attribute?

Reply:
@eliotwrobson From what I'm reading, `@property` is read-only by default. I'd have to declare `@input_parameters.setter` on a new method to enable the setter. But I also tested in the REPL, and our `__setattr__` override kicks in anyway.

```
>>> dfa.input_parameters = ''
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/caleb/Repositories/personal/automata/automata/base/automaton.py", line 30, in __setattr__
    raise AttributeError(f'This {type(self).__name__} is immutable')
AttributeError: This DFA is immutable
```

Sources:
https://www.programiz.com/python-programming/property
https://docs.python.org/3/library/functions.html